// angular2-notifier/src/services/notifier.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var view_container_ref_1 = require('@angular/core/src/linker/view_container_ref');
var notifier_notification_model_1 = require('./../models/notifier-notification.model');
var notifier_action_model_1 = require('./../models/notifier-action.model');
var notifier_container_component_1 = require('./../components/notifier-container.component');
var NotifierService = (function () {
    function NotifierService(applicationRef, componentResolver) {
        var _this = this;
        this.componentResolver = componentResolver;
        if (applicationRef['_rootComponents'].length === 0) {
            applicationRef.registerBootstrapListener(function (rootComponentRef) {
                _this.setupComponent(rootComponentRef);
            });
        }
        else {
            this.setupComponent(applicationRef['_rootComponents'][0]);
        }
    }
    NotifierService.prototype.notify = function (type, message) {
        return this.notifierContainer.doAction({
            payload: new notifier_notification_model_1.NotifierNotification(type, message),
            type: notifier_action_model_1.SHOW
        });
    };
    NotifierService.prototype.default = function (message) {
        return this.notify('default', message);
    };
    NotifierService.prototype.info = function (message) {
        return this.notify('info', message);
    };
    NotifierService.prototype.success = function (message) {
        return this.notify('success', message);
    };
    NotifierService.prototype.warning = function (message) {
        return this.notify('warning', message);
    };
    NotifierService.prototype.error = function (message) {
        return this.notify('error', message);
    };
    NotifierService.prototype.clearAll = function () {
        return this.notifierContainer.doAction({
            type: notifier_action_model_1.CLEAR_ALL
        });
    };
    NotifierService.prototype.clearOldest = function () {
        return this.notifierContainer.doAction({
            type: notifier_action_model_1.CLEAR_OLDEST
        });
    };
    NotifierService.prototype.clearNewest = function () {
        return this.notifierContainer.doAction({
            type: notifier_action_model_1.CLEAR_NEWEST
        });
    };
    NotifierService.prototype.setupComponent = function (rootComponent) {
        var _this = this;
        var rootContainer = new view_container_ref_1.ViewContainerRef_(rootComponent['_hostElement']);
        this.componentResolver
            .resolveComponent(notifier_container_component_1.NotifierContainerComponent)
            .then(function (componentFactory) {
            _this.notifierContainer = rootContainer
                .createComponent(componentFactory, rootContainer.length, rootContainer.parentInjector)
                .instance;
        });
    };
    NotifierService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [core_1.ApplicationRef, core_1.ComponentResolver])
    ], NotifierService);
    return NotifierService;
}());
exports.NotifierService = NotifierService;
