// angular2-notifier/src/services/notifier-animations.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var NotifierAnimationService = (function () {
    function NotifierAnimationService() {
        this.setupAnimationPresets();
    }
    NotifierAnimationService.prototype.getAnimation = function (name, way, component) {
        var keyframes = this.animationPresets[name][way](component);
        var duration = (way === 'in')
            ? component.getConfig().animations.show.duration
            : component.getConfig().animations.hide.duration;
        var easing = (way === 'in')
            ? component.getConfig().animations.show.easing
            : component.getConfig().animations.hide.easing;
        return {
            keyframes: [
                keyframes.from,
                keyframes.to
            ],
            options: {
                delay: 10,
                duration: duration,
                easing: easing,
                fill: 'forwards'
            }
        };
    };
    NotifierAnimationService.prototype.setupAnimationPresets = function () {
        this.animationPresets = {
            fade: {
                in: function (component) {
                    return {
                        from: {
                            opacity: 0
                        },
                        to: {
                            opacity: 1
                        }
                    };
                },
                out: function (component) {
                    return {
                        from: {
                            opacity: 1
                        },
                        to: {
                            opacity: 0
                        }
                    };
                }
            },
            slide: {
                in: function (component) {
                    var animationStart;
                    var animationEnd;
                    switch (component.getConfig().position.horizontal.position) {
                        case 'left':
                            var leftPosition = "calc( -100% - " + component.getConfig().position.horizontal.distance + "px - 10px )";
                            animationStart = {
                                transform: "translate3d( " + leftPosition + ", 0, 0 )"
                            };
                            animationEnd = {
                                transform: 'translate3d( 0, 0, 0 )'
                            };
                            break;
                        case 'middle':
                            var middlePosition = void 0;
                            switch (component.getConfig().position.vertical.position) {
                                case 'top':
                                    middlePosition = "calc( -100% - " + component.getConfig().position.horizontal.distance + "px - 10px )";
                                    break;
                                case 'bottom':
                                    middlePosition = "calc( 100% + " + component.getConfig().position.horizontal.distance + "px + 10px )";
                                    break;
                            }
                            animationStart = {
                                transform: "translate3d( -50%, " + middlePosition + ", 0 )"
                            };
                            animationEnd = {
                                transform: 'translate3d( -50%, 0, 0 )'
                            };
                            break;
                        case 'right':
                            var rightPosition = "calc( 100% + " + component.getConfig().position.horizontal.distance + "px + 10px )";
                            animationStart = {
                                transform: "translate3d( " + rightPosition + ", 0, 0 )"
                            };
                            animationEnd = {
                                transform: 'translate3d( 0, 0, 0 )'
                            };
                            break;
                    }
                    return {
                        from: animationStart,
                        to: animationEnd
                    };
                },
                out: function (component) {
                    var animationStart;
                    var animationEnd;
                    switch (component.getConfig().position.horizontal.position) {
                        case 'left':
                            var leftPosition = "calc( -100% - " + component.getConfig().position.horizontal.distance + "px - 10px )";
                            animationStart = {
                                transform: "translate3d( 0, " + component.getCurrentShift() + "px, 0 )"
                            };
                            animationEnd = {
                                transform: "translate3d( " + leftPosition + ", " + component.getCurrentShift() + "px, 0 )"
                            };
                            break;
                        case 'middle':
                            var middlePosition = void 0;
                            switch (component.getConfig().position.vertical.position) {
                                case 'top':
                                    middlePosition = "calc( -100% - " + component.getConfig().position.horizontal.distance + "px - 10px )";
                                    break;
                                case 'bottom':
                                    middlePosition = "calc( 100% + " + component.getConfig().position.horizontal.distance + "px + 10px )";
                                    break;
                            }
                            animationStart = {
                                transform: "translate3d( -50%, " + component.getCurrentShift() + "px, 0 )"
                            };
                            animationEnd = {
                                transform: "translate3d( -50%, " + middlePosition + ", 0 )"
                            };
                            break;
                        case 'right':
                            var rightPosition = "calc( 100% + " + component.getConfig().position.horizontal.distance + "px + 10px )";
                            animationStart = {
                                transform: "translate3d( 0, " + component.getCurrentShift() + "px, 0 )"
                            };
                            animationEnd = {
                                transform: "translate3d( " + rightPosition + ", " + component.getCurrentShift() + "px, 0 )"
                            };
                            break;
                    }
                    return {
                        from: animationStart,
                        to: animationEnd
                    };
                }
            }
        };
    };
    ;
    NotifierAnimationService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], NotifierAnimationService);
    return NotifierAnimationService;
}());
exports.NotifierAnimationService = NotifierAnimationService;
